# MICROSOFT MOVIE STUDIO ANALYSIS

## INTRODUCTION

## Business Problem

### Objectives

To solve the above problem and help in steering Microsoft in the right direction in this endeavour, the following are this project's objectives

<b>Main Objective</b>
> Pass

<b>Specific Objectives</b>
1. To ...
2. To ...
3. To ...

## Experimental Design

The research follows the steps as listed:
1. Introduction and Business Problem
2. Loading Data and Relevant Libraries
3. Data Wrangling and Preparation
4. Data Exploration and Visualization
5. Conclusions
6. Recommendations

## Metric of Success
The project will be considered a success once it yields at least three actionable and data-driven recommendations to Mircosoft which can ensure the company focuses on the right steps while breaking into the Movie Industry

## Data Description

To be described

## Loading Libraries

In [1]:
import pandas as pd
import numpy as np
import sqlite3
import seaborn as sns
from datetime import datetime
from matplotlib import pyplot as plt
%matplotlib inline

## Loading Data

In [2]:
# Creating a connection to database file
conn = sqlite3.connect("./data/im.db")


# Query for all tables
tablesQuery = """
SELECT name FROM sqlite_master
WHERE type='table';
"""

# Executing query
pd.read_sql(tablesQuery, conn)

name
0   movie_basics
1      directors
2      known_for
3     movie_akas
4  movie_ratings
5        persons
6     principals
7        writers

> From the abov, there are 8 tables, investigating them reveals the following:

### DataBase Files

**Writers Table**

In [3]:
table1 = """
Select * FROM writers
LIMIT 10;
"""
pd.read_sql(table1, conn)

movie_id  person_id
0  tt0285252  nm0899854
1  tt0438973  nm0175726
2  tt0438973  nm1802864
3  tt0462036  nm1940585
4  tt0835418  nm0310087
5  tt0835418  nm0841532
6  tt0878654  nm0284943
7  tt0878654  nm0284943
8  tt0878654  nm0284943
9  tt0996958  nm2286991

> The table above is giving information about the writers responsible for various movies using respective IDs.

**Directors Table**

In [4]:
table2 = """
SELECT * FROM directors
LIMIT 10
"""

pd.read_sql(table2, conn)

movie_id  person_id
0  tt0285252  nm0899854
1  tt0462036  nm1940585
2  tt0835418  nm0151540
3  tt0835418  nm0151540
4  tt0878654  nm0089502
5  tt0878654  nm2291498
6  tt0878654  nm2292011
7  tt0879859  nm2416460
8  tt0996958  nm2286991
9  tt0996958  nm2286991

> The table above is giving information about the directors responsible for various movies using respective IDs.

**Movie Basics Table**

In [5]:
table3 = """
SELECT * FROM movie_basics
LIMIT 10;
"""

pd.read_sql(table3, conn)

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   
5  tt0111414                      A Thin Life                 A Thin Life   
6  tt0112502                          Bigfoot                     Bigfoot   
7  tt0137204                  Joe Finds Grace             Joe Finds Grace   
8  tt0139613                       O Silêncio                  O Silêncio   
9  tt0144449            Nema aviona za Zagreb       Nema aviona za Zagreb   

   start_year  runtime_minutes                      genres  
0        2013            175.0          Action,Crime,Drama  
1        2019            114.0             Biography,Drama  
2        2018            122.0                       Drama  
3        2018              NaN                Comedy,Drama  
4        2017             80.0        Comedy,Drama,Fantasy  
5        2018             75.0                      Comedy  
6        2017              NaN             Horror,Thriller  
7        2017             83.0  Adventure,Animation,Comedy  
8        2012              NaN         Documentary,History  
9        2012             82.0                   Biography

> The above table gives information about basic attributes of movies:
* Titles
* Runtime in Minutes
* Genres spanned
* Start/Release Year

**Movie AKAs**

In [6]:
table4 = """
SELECT * FROM movie_akas
LIMIT 10;
"""
pd.read_sql(table4, conn)

movie_id  ordering                                    title region  \
0  tt0369610        10                            Джурасик свят     BG   
1  tt0369610        11                        Jurashikku warudo     JP   
2  tt0369610        12  Jurassic World: O Mundo dos Dinossauros     BR   
3  tt0369610        13                  O Mundo dos Dinossauros     BR   
4  tt0369610        14                           Jurassic World     FR   
5  tt0369610        15                           Jurassic World     GR   
6  tt0369610        16                           Jurassic World     IT   
7  tt0369610        17                            Jurski svijet     HR   
8  tt0369610        18                             Olam ha'Yura     IL   
9  tt0369610        19           Jurassic World: Mundo Jurásico     MX   

  language        types   attributes  is_original_title  
0       bg         None         None                0.0  
1     None  imdbDisplay         None                0.0  
2     None  imdbDisplay         None                0.0  
3     None         None  short title                0.0  
4     None  imdbDisplay         None                0.0  
5     None  imdbDisplay         None                0.0  
6     None  imdbDisplay         None                0.0  
7     None  imdbDisplay         None                0.0  
8       he  imdbDisplay         None                0.0  
9     None  imdbDisplay         None                0.0

> At first glance, the above table has no relevant information to use in the analysis that is not already in other preceeding tables

**Movie Ratings Table**

In [7]:
table5 = """
SELECT * FROM movie_ratings
LIMIT 10;
"""

pd.read_sql(table5, conn)

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21
5   tt1069246            6.2       326
6   tt1094666            7.0      1613
7   tt1130982            6.4       571
8   tt1156528            7.2       265
9   tt1161457            4.2       148

> The above table seems to be conveying the average rating for a movie as well as the number of votes contributing to said rating

**Known For**

In [8]:
table6 = """
SELECT * FROM known_for
LIMIT 10;
"""

pd.read_sql(table6, conn)

person_id   movie_id
0  nm0061671  tt0837562
1  nm0061671  tt2398241
2  nm0061671  tt0844471
3  nm0061671  tt0118553
4  nm0061865  tt0896534
5  nm0061865  tt6791238
6  nm0061865  tt0287072
7  nm0061865  tt1682940
8  nm0062070  tt1470654
9  nm0062070  tt0363631

> Going off on a limb, the data above probably represents actors and which works they are famous for.

**Principals**

In [9]:
table7 = """
SELECT * FROM principals
LIMIT 10;
"""

pd.read_sql(table7, conn)

movie_id  ordering  person_id  category       job  \
0  tt0111414         1  nm0246005     actor      None   
1  tt0111414         2  nm0398271  director      None   
2  tt0111414         3  nm3739909  producer  producer   
3  tt0323808        10  nm0059247    editor      None   
4  tt0323808         1  nm3579312   actress      None   
5  tt0323808         2  nm2694680     actor      None   
6  tt0323808         3  nm0574615     actor      None   
7  tt0323808         4  nm0502652   actress      None   
8  tt0323808         5  nm0362736  director      None   
9  tt0323808         6  nm0811056  producer  producer   

                 characters  
0               ["The Man"]  
1                      None  
2                      None  
3                      None  
4          ["Beth Boothby"]  
5         ["Steve Thomson"]  
6  ["Sir Lachlan Morrison"]  
7   ["Lady Delia Morrison"]  
8                      None  
9                      None

> The above data will not be used in the analysis to follow.

**Persons**

In [10]:
table8 = """
SELECT * FROM persons
LIMIT 10;
"""

pd.read_sql(table8, conn)

person_id       primary_name  birth_year death_year  \
0  nm0061671  Mary Ellen Bauder         NaN       None   
1  nm0061865       Joseph Bauer         NaN       None   
2  nm0062070         Bruce Baum         NaN       None   
3  nm0062195       Axel Baumann         NaN       None   
4  nm0062798        Pete Baxter         NaN       None   
5  nm0062879     Ruel S. Bayani         NaN       None   
6  nm0063198              Bayou         NaN       None   
7  nm0063432      Stevie Be-Zet         NaN       None   
8  nm0063618          Jeff Beal      1963.0       None   
9  nm0063750    Lindsay Beamish         NaN       None   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator  
5         director,production_manager,miscellaneous  
6                                             actor  
7                               composer,soundtrack  
8              composer,music_department,soundtrack  
9                             actress,miscellaneous

> This table details what professions key individuals in the database do and when they were born/laid to rest; it is not relevant to the ensuing analysis.

## Loading Data

### From Database

In [11]:
# Joining relevant information into a single table
movieBasicsDf = pd.read_sql("""
                SELECT primary_title, start_year, runtime_minutes, genres, averagerating, numvotes
                FROM movie_basics
                JOIN movie_ratings
                    USING(movie_id)
                WHERE genres IS NOT NULL;
                """, conn)

print(movieBasicsDf.shape, end="\n\n")
movieBasicsDf.head()

(73052, 6)



primary_title  start_year  runtime_minutes  \
0                        Sunghursh        2013            175.0   
1  One Day Before the Rainy Season        2019            114.0   
2       The Other Side of the Wind        2018            122.0   
3                  Sabse Bada Sukh        2018              NaN   
4         The Wandering Soap Opera        2017             80.0   

                 genres  averagerating  numvotes  
0    Action,Crime,Drama            7.0        77  
1       Biography,Drama            7.2        43  
2                 Drama            6.9      4517  
3          Comedy,Drama            6.1        13  
4  Comedy,Drama,Fantasy            6.5       119

## From Flat Files

In [12]:
!ls ./data

bom.movie_gross.csv
im.db
rt.movie_info.tsv
rt.reviews.tsv
tmdb.movies.csv
tn.movie_budgets.csv


In [13]:
movieGrossDf = pd.read_csv("./data/bom.movie_gross.csv")
movieInfoDf = pd.read_csv("./data/rt.movie_info.tsv", delimiter="\t")
movieBudgetsDf = pd.read_csv("./data/tn.movie_budgets.csv")
imdbRatingsDf = pd.read_csv("./data/tmdb.movies.csv", index_col=0)

### Basic Information from Datasets

In [14]:
# Movie Basics
movieBasicsDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73052 entries, 0 to 73051
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   primary_title    73052 non-null  object 
 1   start_year       73052 non-null  int64  
 2   runtime_minutes  65720 non-null  float64
 3   genres           73052 non-null  object 
 4   averagerating    73052 non-null  float64
 5   numvotes         73052 non-null  int64  
dtypes: float64(2), int64(2), object(2)
memory usage: 3.3+ MB


> The above data has 6 columns and 73,052 rows.
All of the null values are focused around runtime length.
The data types for respective fields is seemingly appropriate. 

In [15]:
# Movie Gross
movieGrossDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


> The above data has 3387 rows and 5 columns.

1. Studio field has 5 null values
2. Domestic gross has 28 null values
3. Foreign gross has 350 null values

The only field with an incorrect data type for analysis purpose is Foreign Gross. This needs to be converted to numeric.

In [16]:
# Movie Info
movieInfoDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1560 entries, 0 to 1559
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            1560 non-null   int64 
 1   synopsis      1498 non-null   object
 2   rating        1557 non-null   object
 3   genre         1552 non-null   object
 4   director      1361 non-null   object
 5   writer        1111 non-null   object
 6   theater_date  1201 non-null   object
 7   dvd_date      1201 non-null   object
 8   currency      340 non-null    object
 9   box_office    340 non-null    object
 10  runtime       1530 non-null   object
 11  studio        494 non-null    object
dtypes: int64(1), object(11)
memory usage: 146.4+ KB


In [17]:
movieInfoDf.head()

id                                           synopsis rating  \
0   1  This gritty, fast-paced, and innovative police...      R   
1   3  New York City, not-too-distant-future: Eric Pa...      R   
2   5  Illeana Douglas delivers a superb performance ...      R   
3   6  Michael Douglas runs afoul of a treacherous su...      R   
4   7                                                NaN     NR   

                                 genre          director  \
0  Action and Adventure|Classics|Drama  William Friedkin   
1    Drama|Science Fiction and Fantasy  David Cronenberg   
2    Drama|Musical and Performing Arts    Allison Anders   
3           Drama|Mystery and Suspense    Barry Levinson   
4                        Drama|Romance    Rodney Bennett   

                            writer  theater_date      dvd_date currency  \
0                   Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
1     David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
2                   Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
3  Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
4                     Giles Cooper           NaN           NaN      NaN   

  box_office      runtime             studio  
0        NaN  104 minutes                NaN  
1    600,000  108 minutes  Entertainment One  
2        NaN  116 minutes                NaN  
3        NaN  128 minutes                NaN  
4        NaN  200 minutes                NaN

> The data has 12 columns and 1560 rows with a lot of null values

The columns with meaningul information to the analysis - rating, director, runtime and studio, need wrangling(Consistency and Validity concerns) to be analysis ready.

In [18]:
# Movie Budgets

movieBudgetsDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


> The above data has 6 columns and 5782 rows with numbers-focused fields being casted as object-types.
There are no missing values.

In [19]:
# IMDB data

imdbRatingsDf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26517 entries, 0 to 26516
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   genre_ids          26517 non-null  object 
 1   id                 26517 non-null  int64  
 2   original_language  26517 non-null  object 
 3   original_title     26517 non-null  object 
 4   popularity         26517 non-null  float64
 5   release_date       26517 non-null  object 
 6   title              26517 non-null  object 
 7   vote_average       26517 non-null  float64
 8   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 2.0+ MB


> The data above has 9 columns and 26517 rows with no missing values.
Release date needs to be cast as datetime

## Data Cleaning

### Movie Basics Dataset

In [20]:
# Preview the first five rows

movieBasicsDf.head()

primary_title  start_year  runtime_minutes  \
0                        Sunghursh        2013            175.0   
1  One Day Before the Rainy Season        2019            114.0   
2       The Other Side of the Wind        2018            122.0   
3                  Sabse Bada Sukh        2018              NaN   
4         The Wandering Soap Opera        2017             80.0   

                 genres  averagerating  numvotes  
0    Action,Crime,Drama            7.0        77  
1       Biography,Drama            7.2        43  
2                 Drama            6.9      4517  
3          Comedy,Drama            6.1        13  
4  Comedy,Drama,Fantasy            6.5       119

#### Completeness

Since the missing values account for roughly 8000/73000 rows, imputing them with aggregate values would skew the data. The best option in this scenario would be to drop them as they would have minimal effect on how much data is left.

In [21]:
# Dropping nulll values

movieBasicsDf.dropna(subset=['runtime_minutes'],inplace=True)

In [22]:
movieBasicsDf.isna().sum()

primary_title      0
start_year         0
runtime_minutes    0
genres             0
averagerating      0
numvotes           0
dtype: int64

#### Consistency

In [23]:
# Checking for duplicates
movieBasicsDf.duplicated(subset=['primary_title']).sum()

3276

In [24]:
# Filtering for duplicates
basicsDuplicates = movieBasicsDf.loc[movieBasicsDf.duplicated(subset=['primary_title'],
                                                             keep=False)]

# Filtering out a `SettingWithCopyWarning`
pd.options.mode.chained_assignment = None
# Sorting by Primary Title and then Start Year
basicsDuplicates.sort_values(by=["primary_title", 'start_year'], inplace=True,
                            ascending=True)

# Printing the shape of the duplicated data
print(basicsDuplicates.shape, end="\n\n")

# Viewing the first 20 records
basicsDuplicates.head(20)

(5666, 6)



primary_title  start_year  runtime_minutes  \
25696             1        2013            112.0   
12137             1        2019            100.0   
12367     100% Love        2011            141.0   
26467     100% Love        2012            166.0   
12001           180        2011            121.0   
14435           180        2011            107.0   
8025           1960        2010             75.0   
33615          1960        2013             66.0   
57234          1985        2016            114.0   
60879          1985        2018             85.0   
38992          1989        2014             54.0   
42650          1989        2014             97.0   
30470       21 Days        2014             89.0   
44378       21 Days        2014             70.0   
3725              3        2010            119.0   
17634             3        2012            145.0   
23394             3        2012            115.0   
54773             3        2018             80.0   
45498           3/4        2014             80.0   
64616           3/4        2017             83.0   

                             genres  averagerating  numvotes  
25696     Documentary,History,Sport            8.0      3452  
12137                         Drama            9.2        13  
12367          Comedy,Drama,Romance            7.0      2331  
26467          Comedy,Drama,Romance            6.0       296  
12001                 Drama,Romance            6.0       688  
14435                  Comedy,Drama            7.4        42  
8025                    Documentary            6.8        49  
33615         Biography,Documentary            5.9         8  
57234        Adventure,Comedy,Drama            8.1        30  
60879                         Drama            6.8       998  
38992                   Documentary            6.8         5  
42650  Documentary,History,Thriller            7.5       175  
30470               Horror,Thriller            4.5       454  
44378                   Documentary            6.0         5  
3725           Comedy,Drama,Romance            6.7      4722  
17634         Drama,Mystery,Romance            7.2      3471  
23394                  Comedy,Drama            6.1       101  
54773       Horror,Mystery,Thriller            3.9       166  
45498                         Drama            5.8         8  
64616                         Drama            6.3       196

>All duplicates will be dropped. Also, note that the above ouptut shows all rows WITH their duplicates due to the argument `keep` being set to True

In [25]:
# Resetting the option to `Warn`
pd.options.mode.chained_assignment = 'warn'

In [26]:
# Dropping the duplicates
movieBasicsDf.drop_duplicates(subset=['primary_title'], keep='first',
                             inplace=True)

#### Validity

In [27]:
# Checking for proper data types within the dataframe
movieBasicsDf.dtypes

primary_title       object
start_year           int64
runtime_minutes    float64
genres              object
averagerating      float64
numvotes             int64
dtype: object

>Everything seems to be in order for this dataframe

### Movie Gross Dataset

In [28]:
movieGrossDf.shape

(3387, 5)

#### Completeness

In [29]:
# Checking for null values
movieGrossDf.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

<p> Nearly half of the rows in this dataset have null values in the _foreign_gross_ column. From speculation, this could be one of two reasons:

1. The data was never recorded, or
2. The movie did not make any money inernationally.

Either way, these will be imputed with a value of 0. The same goes for Domestic Gross
</p>

In [30]:
# Imputing nulls in numeric columns
movieGrossDf['domestic_gross'].fillna(value='0', inplace=True)
movieGrossDf['foreign_gross'].fillna(value='0', inplace=True)

> Please note that attempting to fill the null values in the above cell with 0 instead of `"0"` brings some complucations when filtering for values with commas. This is due to the conflict of data types: imputing numerical values to a column populated with objects

In [31]:
# Dropping nulls in studio column
movieGrossDf.dropna(subset=['studio'], inplace=True)

#### Consistency

In [32]:
# Checking for duplicates
movieGrossDf.duplicated().sum()

0

#### Validity

In [33]:
# Checking column data types
movieGrossDf.dtypes

title             object
studio            object
domestic_gross    object
foreign_gross     object
year               int64
dtype: object

> Foreign gross needs to be numeric

In [34]:
movieGrossDf['foreign_gross'].head(10)

0    652000000
1    691300000
2    664300000
3    535700000
4    513900000
5    398000000
6    311500000
7    391000000
8    291600000
9    277300000
Name: foreign_gross, dtype: object

In [35]:
# Filtering only for values containing a comma as a value separator.
movieGrossDf.loc[movieGrossDf['foreign_gross'].str.contains(',')]

title studio domestic_gross foreign_gross  year
1872  Star Wars: The Force Awakens     BV    936700000.0       1,131.6  2015
1873                Jurassic World   Uni.    652300000.0       1,019.4  2015
1874                     Furious 7   Uni.    353000000.0       1,163.0  2015
2760       The Fate of the Furious   Uni.    226000000.0       1,010.0  2017
3079        Avengers: Infinity War     BV    678800000.0       1,369.5  2018

> Before converting the column's information to numerical values, these commas must be dealt with. It is wprth noting that these numbers should also be converted into millions

In [36]:
# Removing the commas

movieGrossDf['foreign_gross'] = movieGrossDf['foreign_gross'].str.replace(',', '')


# Converting column data into numerical data
movieGrossDf['foreign_gross'] = movieGrossDf['foreign_gross'].astype('float')
movieGrossDf['domestic_gross'] = movieGrossDf['domestic_gross'].astype('float')

In [118]:
billions_movies_indices = [1872, 1873, 1874, 2760, 3079]

for idx in billions_movies_indices:
    value = movieGrossDf.at[idx, 'foreign_gross']
    
    # Updating the record
    movieGrossDf.at[idx, 'foreign_gross'] = value * 10**6
    
    
# Displaying one of the results

movieGrossDf.loc[1872,:]

title             Star Wars: The Force Awakens
studio                                      BV
domestic_gross                     936700000.0
foreign_gross                     1131600000.0
year                                      2015
Name: 1872, dtype: object

### Movie Info Dataframe

In [37]:
movieInfoDf.head()

id                                           synopsis rating  \
0   1  This gritty, fast-paced, and innovative police...      R   
1   3  New York City, not-too-distant-future: Eric Pa...      R   
2   5  Illeana Douglas delivers a superb performance ...      R   
3   6  Michael Douglas runs afoul of a treacherous su...      R   
4   7                                                NaN     NR   

                                 genre          director  \
0  Action and Adventure|Classics|Drama  William Friedkin   
1    Drama|Science Fiction and Fantasy  David Cronenberg   
2    Drama|Musical and Performing Arts    Allison Anders   
3           Drama|Mystery and Suspense    Barry Levinson   
4                        Drama|Romance    Rodney Bennett   

                            writer  theater_date      dvd_date currency  \
0                   Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
1     David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
2                   Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
3  Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
4                     Giles Cooper           NaN           NaN      NaN   

  box_office      runtime             studio  
0        NaN  104 minutes                NaN  
1    600,000  108 minutes  Entertainment One  
2        NaN  116 minutes                NaN  
3        NaN  128 minutes                NaN  
4        NaN  200 minutes                NaN

#### Validity

In [38]:
movieInfoDf.dtypes

id               int64
synopsis        object
rating          object
genre           object
director        object
writer          object
theater_date    object
dvd_date        object
currency        object
box_office      object
runtime         object
studio          object
dtype: object

In [39]:
# Converting dates into datetime formats

movieInfoDf['theater_date'] = pd.to_datetime(movieInfoDf['theater_date'], 
                                             infer_datetime_format=True)
movieInfoDf['dvd_date'] = pd.to_datetime(movieInfoDf['dvd_date'], 
                                             infer_datetime_format=True)

In [40]:
# Types of currencies in the currency column

movieInfoDf['currency'].value_counts()

$    340
Name: currency, dtype: int64

In [41]:
# Considering there is another dataset dedicated to the money the movies
# made, this is redundant information

movieInfoDf.drop(labels=['currency'], axis=1, inplace=True)

In [42]:
# Fixing the runtime column

# Drop minutes text

movieInfoDf['runtime'] = movieInfoDf['runtime'].str.replace(' minutes', '')

#### Completeness

In [44]:
movieInfoDf.isnull().sum()

id                 0
synopsis          62
rating             3
genre              8
director         199
writer           449
theater_date     359
dvd_date         359
box_office      1220
runtime           30
studio          1066
dtype: int64

In [46]:
# Threshold of five null values qualifies the row for dropping
movieInfoDf.dropna(axis=0, thresh=5, inplace=True)

In [47]:
movieInfoDf.loc[movieInfoDf['runtime'].isnull()]

id                                           synopsis rating  \
11      18  In 1979, Bill Viola and Frank Caliguri dreamed...     NR   
102    131  No Sesame. All Street. THE HAPPYTIME MURDERS i...      R   
434    567  Now graduated from college and out in the real...  PG-13   
516    676  The Hill would have made a terrific Samuel Ful...     NR   
536    699                                                NaN     NR   
579    749  Jalaibee is a Tale of two friends Billu &amp; ...     NR   
750    968  Arnab a typical Bengali man is making his way ...     NR   
976   1267                                                NaN     NR   
1023  1325  From the outer reaches of space to the small-t...      R   
1078  1389  From Ron Shelton, writer/director of Tin Cup a...  PG-13   
1126  1451  A group of scientific researchers on a space s...     NR   
1201  1541  In the heatwarming live action adventure "Disn...     PG   
1499  1931  Mark Felt - The Man Who Brought Down the White...  PG-13   

                                                  genre         director  \
11                                          Documentary              NaN   
102                         Action and Adventure|Comedy     Brian Henson   
434                                              Comedy        Trish Sie   
516                          Action and Adventure|Drama    Robert Iscove   
536                                               Drama     Chris Menges   
579   Action and Adventure|Art House and Internation...     Yasir Jaswal   
750   Art House and International|Horror|Mystery and...       Prosit Roy   
976   Action and Adventure|Kids and Family|Science F...      Guy Ritchie   
1023  Action and Adventure|Horror|Science Fiction an...      Shane Black   
1078                        Action and Adventure|Comedy      Ron Shelton   
1126   Mystery and Suspense|Science Fiction and Fantasy      Julius Onah   
1201        Action and Adventure|Comedy|Kids and Family     Marc Forster   
1499                                              Drama  Peter Landesman   

                                          writer theater_date   dvd_date  \
11                                  Robert Zullo          NaT        NaT   
102             Todd Berger|Dee Austin Robertson   2018-08-24 2018-12-04   
434                        Mike White|Kay Cannon   2017-12-22 2018-03-20   
516                                          NaN          NaT        NaT   
536                                  Shawn Slovo          NaT        NaT   
579                                          NaN          NaT        NaT   
750   Abhishek Bannerjee|Prosit Roy|Rajat Kapoor          NaT        NaT   
976                      John August|Guy Ritchie          NaT        NaT   
1023                     Shane Black|Fred Dekker   2018-09-14 2018-11-27   
1078                                 Ron Shelton   2017-12-08 2018-02-27   
1126                                  Oren Uziel   2018-02-04 2019-02-05   
1201                             Alex Ross Perry   2018-08-03 2018-11-06   
1499                             Peter Landesman   2017-09-29 2018-01-09   

       box_office runtime                      studio  
11            NaN     NaN  Showtime Documentary Films  
102           NaN     NaN                    STXfilms  
434   104,880,310     NaN                         NaN  
516           NaN     NaN                         NaN  
536           NaN     NaN                         NaN  
579           NaN     NaN                         NaN  
750           NaN     NaN                         NaN  
976           NaN     NaN                         NaN  
1023          NaN     NaN                         NaN  
1078          NaN     NaN        Broad Green Pictures  
1126          NaN     NaN                         NaN  
1201          NaN     NaN        Walt Disney Pictures  
1499      766,428     NaN      Sony Pictures Classics

> Runtime and Studio are important features in this analysis and having null values makes the quality of work subpar.

In [48]:
movieInfoDf.dropna(subset=['studio', 'runtime'], inplace=True)

In [49]:
movieInfoDf.isna().sum()

id                0
synopsis          1
rating            0
genre             0
director         53
writer          100
theater_date     29
dvd_date         29
box_office      185
runtime           0
studio            0
dtype: int64

* Impute writer and director columns with  `unknown`
* Impute box office with 0.
* Drop rows without date information

In [52]:
movieInfoDf['writer'].fillna(value="unknown", inplace=True)
movieInfoDf[ 'director'].fillna(value="unknown", inplace=True)
movieInfoDf.dropna(subset=['theater_date', 'dvd_date'], axis=0, inplace=True)
movieInfoDf['box_office'].fillna(value=0, inplace=True)

In [53]:
movieInfoDf.isnull().sum()

id              0
synopsis        0
rating          0
genre           0
director        0
writer          0
theater_date    0
dvd_date        0
box_office      0
runtime         0
studio          0
dtype: int64

#### Consistency

In [55]:
# Checking for duplicated data
movieInfoDf.duplicated(subset=['id']).sum()

0

### Movie Budgets Dataset

#### Completeness

In [56]:
movieBudgetsDf.isnull().sum()

id                   0
release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
dtype: int64

> No null values present in the dataframe

#### Consistency

In [57]:
movieBudgetsDf.duplicated(subset=['id']).sum()

5682

In [67]:
# Investigating the numerous duplicates based on the unique identifier `id`
movieBudgetsDf.loc[movieBudgetsDf.duplicated(subset=['id'])]

id  release_date                          movie production_budget  \
100    1  May 29, 2009                             Up      $175,000,000   
101    2  Nov 22, 2017                           Coco      $175,000,000   
102    3  Nov 21, 2018      Ralph Breaks The Internet      $175,000,000   
103    4  Mar 27, 2009            Monsters vs. Aliens      $175,000,000   
104    5  Jun 16, 2017                         Cars 3      $175,000,000   
...   ..           ...                            ...               ...   
5777  78  Dec 31, 2018                         Red 11            $7,000   
5778  79   Apr 2, 1999                      Following            $6,000   
5779  80  Jul 13, 2005  Return to the Land of Wonders            $5,000   
5780  81  Sep 29, 2015           A Plague So Pleasant            $1,400   
5781  82   Aug 5, 2005              My Date With Drew            $1,100   

     domestic_gross worldwide_gross  
100    $293,004,164    $731,463,377  
101    $209,726,015    $798,008,101  
102    $201,091,711    $524,283,695  
103    $198,351,526    $381,687,380  
104    $152,901,115    $383,541,369  
...             ...             ...  
5777             $0              $0  
5778        $48,482        $240,495  
5779         $1,338          $1,338  
5780             $0              $0  
5781       $181,041        $181,041  

[5682 rows x 6 columns]

In [68]:
# A closer look reveals the following

movieBudgetsDf.loc[movieBudgetsDf['id']==1]

id  release_date                                 movie  \
0      1  Dec 18, 2009                                Avatar   
100    1  May 29, 2009                                    Up   
200    1   Mar 7, 2014                 Mr. Peabody & Sherman   
300    1  Dec 17, 2010                      How Do You Know?   
400    1  Dec 11, 2015               In the Heart of the Sea   
500    1   Jun 8, 2007                      Ocean's Thirteen   
600    1  Nov 26, 2003                              Timeline   
700    1  Aug 21, 2009                  Inglourious Basterds   
800    1  Nov 21, 2012                              Red Dawn   
900    1  Nov 26, 1997                   Alien: Resurrection   
1000   1  May 12, 1995                          Crimson Tide   
1100   1  Dec 19, 2008                               Yes Man   
1200   1   Mar 3, 2000             What Planet Are You From?   
1300   1  Dec 26, 2008                    Revolutionary Road   
1400   1  Nov 14, 2014                    Dumb and Dumber To   
1500   1   Mar 9, 2012                      A Thousand Words   
1600   1  Jul 18, 2008                          Space Chimps   
1700   1  Mar 26, 2004                           Jersey Girl   
1800   1  Feb 18, 2011     Big Mommas: Like Father, Like Son   
1900   1  Sep 22, 1989                            Black Rain   
2000   1  Jun 15, 2007                    DOA: Dead or Alive   
2100   1  Jul 16, 1999                           Lake Placid   
2200   1  Sep 19, 1990                            Goodfellas   
2300   1   Dec 7, 2005            The World's Fastest Indian   
2400   1  Jun 24, 1987                            Spaceballs   
2500   1  Jun 16, 1978                                Jaws 2   
2600   1  Sep 21, 2007                         Into the Wild   
2700   1  Dec 10, 2010                           The Tempest   
2800   1  Nov 14, 1980                           Raging Bull   
2900   1  Apr 21, 2006                       American Dreamz   
3000   1  Dec 20, 1996                                Scream   
3100   1  Apr 23, 1993                         The Dark Half   
3200   1  Aug 21, 1998                    Dead Man on Campus   
3300   1  Feb 26, 2010                          Un Prophète   
3400   1  Mar 20, 1981        The Postman Always Rings Twice   
3500   1   Feb 1, 2002                              Slackers   
3600   1  Sep 16, 2016                           Blair Witch   
3700   1  Oct 11, 2002                            Swept Away   
3800   1  Sep 29, 2000                             Beautiful   
3900   1   Oct 8, 2010            It's Kind of a Funny Story   
4000   1   Aug 4, 2006                           The Descent   
4100   1  Nov 13, 2009                         The Messenger   
4200   1  Mar 18, 2011                     Winter in Wartime   
4300   1  Apr 10, 1981                            Nighthawks   
4400   1  Jan 23, 2015                                 Mommy   
4500   1   Mar 1, 2000                             3 Strikes   
4600   1  Jul 18, 1959                       The Nun's Story   
4700   1  Jul 31, 2009                         The Collector   
4800   1   Feb 7, 1974                       Blazing Saddles   
4900   1  Jun 13, 1962                                Lolita   
5000   1   Dec 1, 2015                            Shark Lake   
5100   1  Mar 22, 2002                         Stolen Summer   
5200   1  May 14, 2003  Dracula: Pages from a Virgin's Diary   
5300   1  Dec 31, 2015                     Karachi se Lahore   
5400   1  Dec 14, 2018                 That Way Madness Lies   
5500   1   Jul 1, 2014                        Mutual Friends   
5600   1  Feb 24, 2015                       Give Me Shelter   
5700   1  May 30, 2008                     The Foot Fist Way   

     production_budget domestic_gross worldwide_gross  
0         $425,000,000   $760,507,625  $2,776,345,279  
100       $175,000,000   $293,004,164    $731,463,377  
200       $145,000,000   $111,506,430    $269,806,430  
300  

> `ID` does not uniquely identify records and is therefore useless. Moreover, checking duplicates should be on the basis of the movie title

In [70]:
movieBudgetsDf.duplicated(subset=['movie']).sum()

84

In [75]:
movieBudgetsDf.loc[movieBudgetsDf.duplicated(subset=['movie'],
                                            keep=False)].sort_values(by="movie").head(20)

id  release_date                         movie production_budget  \
4270  71  Dec 23, 1954  20,000 Leagues Under the Sea        $5,000,000   
5614  15  Dec 24, 1916  20,000 Leagues Under the Sea          $200,000   
1648  49  Apr 30, 2010     A Nightmare on Elm Street       $35,000,000   
5016  17   Nov 9, 1984     A Nightmare on Elm Street        $1,800,000   
2032  33  Nov 11, 1992                       Aladdin       $28,000,000   
80    81  May 24, 2019                       Aladdin      $182,000,000   
50    51   Mar 5, 2010           Alice in Wonderland      $200,000,000   
4759  60  Jul 28, 1951           Alice in Wonderland        $3,000,000   
4120  21  Oct 17, 1956   Around the World in 80 Days        $6,000,000   
340   41  Jun 16, 2004   Around the World in 80 Days      $110,000,000   
134   35  Mar 17, 2017          Beauty and the Beast      $160,000,000   
2485  86  Nov 13, 1991          Beauty and the Beast       $20,000,000   
3007   8  Nov 18, 1959                       Ben-Hur       $15,000,000   
422   23  Aug 19, 2016                       Ben-Hur       $95,000,000   
3307   8  Aug 14, 2015                      Brothers       $13,000,000   
2140  41   Dec 4, 2009                      Brothers       $26,000,000   
4058  59  Mar 26, 2010                          Ca$h        $7,000,000   
3231  32  Dec 31, 2008                          Ca$h       $13,700,000   
1925  26  Oct 18, 2013                        Carrie       $30,000,000   
5015  16  Nov 16, 1976                        Carrie        $1,800,000   

     domestic_gross worldwide_gross  
4270    $28,200,000     $28,200,000  
5614     $8,000,000      $8,000,000  
1648    $63,075,011    $117,729,621  
5016    $25,504,513     $25,504,513  
2032   $217,350,219    $504,050,219  
80     $246,734,314    $619,234,314  
50     $334,191,110  $1,025,491,110  
4759             $0              $0  
4120    $42,000,000     $42,000,000  
340     $24,004,159     $72,004,159  
134    $504,014,165  $1,259,199,706  
2485   $376,057,266    $608,431,132  
3007    $73,000,000     $73,259,017  
422     $26,410,477     $91,669,648  
3307       $656,688     $17,856,688  
2140    $28,544,157     $45,043,870  
4058        $46,488         $46,488  
3231             $0     $11,738,256  
1925    $35,266,619     $82,409,520  
5015    $25,878,153     $25,878,153

> The above subset of data raises concerns on which record should be kept. Investigating a more popular movie like The Avengers will serve as a good guide on which occurence of a duplicated record will be kept.

In [76]:
movieBudgetsDf.loc[movieBudgetsDf['movie']=='The Avengers']

id  release_date         movie production_budget domestic_gross  \
26   27   May 4, 2012  The Avengers      $225,000,000   $623,279,547   
934  35  Aug 14, 1998  The Avengers       $60,000,000    $23,385,416   

    worldwide_gross  
26   $1,517,935,897  
934     $48,585,416

> Perhaps these aren't duplicates but rather movies and their remakes given they have different release dates.

> This site [here](https://m.the-numbers.com/movie/Avengers-The-(2012)#tab=sumary) confirms the actual numbers that this movie(2012) version made. The other movie is also an actual movie with actual records in place

In [77]:
# Improving the search criteria for duplicates

movieBudgetsDf.duplicated(subset=['release_date', 'movie']).sum()

0

#### Validity

> The id column may uniquely identify these movies in ANOTHER table and therefore cannot be dropped

In [78]:
# Data types in specific fields need to be cleaned
movieBudgetsDf.dtypes

id                    int64
release_date         object
movie                object
production_budget    object
domestic_gross       object
worldwide_gross      object
dtype: object

In [81]:
# Date column

movieBudgetsDf['release_date'] = pd.to_datetime(movieBudgetsDf['release_date'])

In [94]:
# Currency columns

movieBudgetsDf['production_budget'] = movieBudgetsDf['production_budget'].str.replace('$', '', regex=False).str.replace(',', '')
movieBudgetsDf['domestic_gross'] = movieBudgetsDf['domestic_gross'].str.replace('$', '', regex=False).str.replace(',', '')
movieBudgetsDf['worldwide_gross'] = movieBudgetsDf['worldwide_gross'].str.replace('$', '', regex=False).str.replace(',', '')

In [95]:
# Casting as appropriate dtypes
movieBudgetsDf['production_budget']=movieBudgetsDf['production_budget'].astype('int64')
movieBudgetsDf['domestic_gross']=movieBudgetsDf['domestic_gross'].astype('int64')
movieBudgetsDf['worldwide_gross']=movieBudgetsDf['worldwide_gross'].astype('int64')

In [96]:
movieBudgetsDf.dtypes

id                            int64
release_date         datetime64[ns]
movie                        object
production_budget             int64
domestic_gross                int64
worldwide_gross               int64
dtype: object

### Imdb Ratings Dataset

In [97]:
imdbRatingsDf.head()

genre_ids     id original_language  \
0      [12, 14, 10751]  12444                en   
1  [14, 12, 16, 10751]  10191                en   
2        [12, 28, 878]  10138                en   
3      [16, 35, 10751]    862                en   
4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

#### Completeness

In [98]:
imdbRatingsDf.isnull().sum()

genre_ids            0
id                   0
original_language    0
original_title       0
popularity           0
release_date         0
title                0
vote_average         0
vote_count           0
dtype: int64

#### Consistency

In [100]:
imdbRatingsDf.duplicated(subset=['id']).sum()

1020

In [102]:
# First few records with their duplicates

imdbRatingsDf.loc[imdbRatingsDf.duplicated(subset=['id'], keep=False)].sort_values(by='id').head(10)

genre_ids   id original_language              original_title  \
20626  [16, 10751, 14]  129                ja                    千と千尋の神隠し   
14173  [16, 10751, 14]  129                ja                    千と千尋の神隠し   
43         [35, 10749]  239                en            Some Like It Hot   
24000      [35, 10749]  239                en            Some Like It Hot   
20639    [28, 53, 878]  280                en  Terminator 2: Judgment Day   
17395    [28, 53, 878]  280                en  Terminator 2: Judgment Day   
14222  [18, 36, 10752]  387                de                    Das Boot   
2494   [18, 36, 10752]  387                de                    Das Boot   
3      [16, 35, 10751]  862                en                   Toy Story   
2473   [16, 35, 10751]  862                en                   Toy Story   

       popularity release_date                       title  vote_average  \
20626      32.043   2002-09-20               Spirited Away           8.5   
14173      32.043   2002-09-20               Spirited Away           8.5   
43         14.200   1959-03-18            Some Like It Hot           8.2   
24000      14.200   1959-03-18            Some Like It Hot           8.2   
20639      24.604   1991-07-03  Terminator 2: Judgment Day           7.9   
17395      24.604   1991-07-03  Terminator 2: Judgment Day           7.9   
14222      16.554   1982-02-10                    Das Boot           8.1   
2494       16.554   1982-02-10                    Das Boot           8.1   
3          28.005   1995-11-22                   Toy Story           7.9   
2473       28.005   1995-11-22                   Toy Story           7.9   

       vote_count  
20626        7424  
14173        7424  
43           1562  
24000        1562  
20639        6682  
17395        6682  
14222         981  
2494          981  
3           10174  
2473        10174

In [103]:
# Dropping duplicates with the a subset in place

imdbRatingsDf.drop_duplicates(subset=['id', 'popularity', 'title'],
                             keep='first', inplace=True)

#### Validity

In [104]:
# Data types present and appropriateness check
imdbRatingsDf.dtypes

genre_ids             object
id                     int64
original_language     object
original_title        object
popularity           float64
release_date          object
title                 object
vote_average         float64
vote_count             int64
dtype: object

In [106]:
# Date column conversion

imdbRatingsDf['release_date'] = pd.to_datetime(imdbRatingsDf['release_date'])

## Data Consolidation